OpenRA.Mods.Cnc/Widgets/Logic/CampaignWorldLogic.cs

```diff
@@ -0,0 +1,624 @@
+                }
+            }
+
+            countryDescriptionTextHeader = WidgetUtils.WrapText(countryDescriptionTextHeader, countryDescriptionHeader.Bounds.Width, countryDescriptionFont);
+            countryDescriptionHeader.Text = countryDescriptionTextHeader;
+            countryDescriptionHeader.Bounds.Height = countryDescriptionFont.Measure(countryDescriptionTextHeader).Y;
+            countryDescriptionTextValues = WidgetUtils.WrapText(countryDescriptionTextValues, countryDescriptionValues.Bounds.Width, countryDescriptionFont);
+            countryDescriptionValues.Text = countryDescriptionTextValues;
+            countryDescriptionValues.Bounds.Height = countryDescriptionFont.Measure(countryDescriptionTextValues).Y;
+            countryDescriptionPanel.ScrollToTop();
+            countryDescriptionPanel.Layout.AdjustChildren();
+
+            selectedMapPreview = Game.ModData.MapCache[nextMap.Uid];
+        }
+
+        void SetPreviewContent()
+        {
+            if (nextMap.Container.Exists("preview.png"))
+                using (var dataStream = nextMap.Container.GetContent("preview.png"))
+                {
+                    nextMap.CustomPreview = new System.Drawing.Bitmap(dataStream);
+                    var preview = new MapPreview(nextMap.Uid, Game.ModData.MapCache);
+                    preview.SetMinimap(new SheetBuilder(SheetType.BGRA).Add(nextMap.CustomPreview));
+                    campaignPreviewWidget.Preview = () => preview;
+                }
+        }
+
+        void PlayAndStart()
+        {
+            try
+            {
+                if (nextMap.PreviewVideo != null)
+                {
+                    videoPlayer.Load(nextMap.PreviewVideo);
+                    videoPlayer.PlayThen(StopVideoAndStart);
+                }
+                else
+                {
+                    StartMission();
+                }
+            }
+            catch (FileNotFoundException)
+            {
+                StartMission();
+            }
+        }
+
+        void StopVideoAndStart()
+        {
+            if (videoStopped)
+            {
+                Sound.MusicVolume = cachedMusicVolume;
+                videoPlayer.Stop();
+            }
+            else
+            {
+                videoStopped = true;
+                StopVideoAndStart();
+                StartMission();
+            }
+        }
+
+        void StartMission()
+        {
+            OrderManager om = null;
+
+            Action lobbyReady = null;
+            lobbyReady = () =>
+            {
+                Game.LobbyInfoChanged -= lobbyReady;
+                onStart();
+                om.IssueOrder(Order.Command("state {0}".F(Session.ClientState.Ready)));
+            };
+            Game.LobbyInfoChanged += lobbyReady;
+
+            // Set Flag if GDI Path to Belarus is selected: gdi04b
+            if (Campaign.Equals("GDI Campaign"))
+            {
+                if (progress == 3 && mapIndex == 1)
```

Comment:
The hack is removed. If you like, you can view the changes in the branch 'nextMission' of our repo.

Best regards
Heiko